lib/runtime-core/src/table/mod.rs

```diff
@@ -18,12 +19,76 @@ pub use self::anyfunc::Anyfunc;
 pub(crate) use self::anyfunc::AnyfuncTable;
 use crate::error::GrowError;
 
+/// Error type indicating why a table access failed.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TableAccessError {
+    /// The index wasn't valid, so no element could be accessed.
+    IndexError,
+
+    // we'll need this error when we support tables holding more types
+    #[allow(dead_code)]
+    /// The type of the table was incorrect, so no element could be accessed.
+    TypeError,
+}
+
+/// Trait indicates types that can be stored in tables
+pub trait StorableInTable: Sized {
+    /// Attempt to lookup self in the given table.
+    fn unwrap_self(storage: &TableStorage, index: u32) -> Result<Self, TableAccessError>;
+
+    /// Wrap value to be stored in a table.
+    fn wrap_self(self, storage: &mut TableStorage, index: u32) -> Result<(), TableAccessError>;
+}
+
+impl<'a, F: Into<Anyfunc<'a>> + TryFrom<Anyfunc<'a>>> StorableInTable for F {
+    fn unwrap_self(storage: &TableStorage, index: u32) -> Result<Self, TableAccessError> {
+        match storage {
+            TableStorage::Anyfunc(ref anyfunc_table) => {
+                let anyfunc = anyfunc_table
+                    .get(index)
+                    .ok_or(TableAccessError::IndexError)?;
+                // Should this be a different error value because it's not a table type error?
+                F::try_from(anyfunc).map_err(|_| TableAccessError::TypeError)
+            }
+        }
+    }
+
+    fn wrap_self(self, storage: &mut TableStorage, index: u32) -> Result<(), TableAccessError> {
+        let anyfunc: Anyfunc = self.into();
+
+        match storage {
+            TableStorage::Anyfunc(ref mut anyfunc_table) => anyfunc_table
+                .set(index, anyfunc)
+                .map_err(|_| TableAccessError::IndexError),
+        }
+    }
+}
+
 /// Kind of table element.
+// note to implementors: all types in `Element` should implement `StorableInTable`.
 pub enum Element<'a> {
     /// Anyfunc.
     Anyfunc(Anyfunc<'a>),
 }
 
+// delegation implementation for `Element`
+impl<'a> StorableInTable for Element<'a> {
```

Comment:
Looking at the reference WebAssembly, Element should tend to be just a `Value` type.

Reply:
Hmm, I was under the impression that that wasn't in the current version of Wasm -- I'm not really following, can you link to the location in the spec you're talking about?

Reply:
https://github.com/WebAssembly/wasm-c-api/blob/master/include/wasm.hh#L344

Reply:
Do we want to expose functionality in our API that is not yet implemented?  Here's the JS API as a counter point: https://webassembly.github.io/spec/js-api/#enumdef-tablekind